patroni/config_generator.py

```diff
@@ -0,0 +1,266 @@
+
+from typing import Any, Dict, Optional
+
+from .config import Config
+from .exceptions import PatroniException
+from .postgresql.config import ConfigHandler
+from .postgresql.misc import postgres_major_version_to_int
+from .utils import get_major_version, patch_config
+
+
+def get_ip() -> str:
+    hostname = None
+    try:
+        hostname = socket.gethostname()
+        return sorted(socket.getaddrinfo(hostname, 0, socket.AF_UNSPEC, socket.SOCK_STREAM, 0),
+                      key=lambda x: x[0])[0][4][0]
+    except OSError as e:
+        sys.exit(f'Failed to define ip address: {e}')
+    except IndexError:
+        sys.exit(f'Failed to define ip address. No address returned by getaddrinfo for {hostname}')
+
+
+def get_bin_dir_from_running_instance(data_dir: str) -> str:
+    postmaster_pid = None
+    try:
+        with open(f"{data_dir}/postmaster.pid", 'r') as f:
+            postmaster_pid = f.readline()
+            if not postmaster_pid:
+                sys.exit('Failed to obtain postmaster pid from postmaster.pid file')
+            postmaster_pid = int(postmaster_pid.strip())
+    except OSError as e:
+        sys.exit(f'Error while reading postmaster.pid file: {e}')
+    import psutil
+    try:
+        return os.path.dirname(psutil.Process(postmaster_pid).exe())
+    except psutil.NoSuchProcess:
+        sys.exit('Obtained postmaster pid doesn\'t exist')
+
+
+def enrich_config_from_running_instance(config: Dict[str, Any], no_value_msg: str, dsn: Optional[str] = None) -> None:
+    """Get
+    - non-internal GUC values having configuration file, postmaster command line or environment variable as a source
+    - postgresql.connect_address, postgresql.listen,
+    - postgresql.pg_hba and postgresql.pg_ident if hba_file/ident_file is set to the default value
+    - superuser auth parameters (from the options used for connection)
+    And redefine scope with the clister_name GUC value if set
+
+    :param config: configuration parameters dict to be enriched
+    :param no_value_msg: str value to be used when a parameter value is not available
+    :param dsn: optional DSN string for the source running instance
+    """
+    from getpass import getuser, getpass
+    from patroni.postgresql.config import parse_dsn
+    from patroni.config import AUTH_ALLOWED_PARAMETERS_MAPPING
+
+    su_params: Dict[str, str] = {}
+    parsed_dsn = {}
+
+    if dsn:
+        parsed_dsn = parse_dsn(dsn)
+        if not parsed_dsn:
+            sys.exit('Failed to parse DSN string')
+
+    # gather auth parameters for the superuser config
+    for conn_param, env_var in AUTH_ALLOWED_PARAMETERS_MAPPING.items():
+        val = parsed_dsn.get(conn_param, os.getenv(env_var))
+        if val:
+            su_params[conn_param] = val
+    # because we use "username" in the config for some reason
+    su_params['username'] = su_params.pop('user', getuser())
+    su_params['password'] = su_params.get('password') or getpass('Please enter the user password:')
```

Comment:
Since we already automatically support all `PATRONI_*` environment variables should we also take into account that the user might have configured `PATRONI_SUPERUSER_*`?